File: PSME/application/src/rest/endpoints/ethernet/ethernet_switch_port_metrics.cpp
/*!
 * @copyright
 * Copyright (c) 2017-2019 Intel Corporation
 *
 * @copyright
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * @copyright
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * @copyright
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * */

#include "psme/rest/constants/constants.hpp"
#include "psme/rest/endpoints/ethernet/ethernet_switch_port_metrics.hpp"



using namespace psme::rest;
using namespace psme::rest::constants;
using namespace agent_framework::module;

namespace {
json::Json make_prototype() {
    json::Json r(json::Json::value_t::object);

    r[Common::ODATA_CONTEXT] = "/redfish/v1/$metadata#EthernetSwitchPortMetrics.EthernetSwitchPortMetrics";
    r[Common::ODATA_ID] = json::Json::value_t::null;
    r[Common::ODATA_TYPE] = "#EthernetSwitchPortMetrics.v1_0_0.EthernetSwitchPortMetrics";
    r[Common::ID] = json::Json::value_t::null;
    r[Common::NAME] = "Ethernet Switch Port Metrics";
    r[Common::DESCRIPTION] = "Ethernet Switch Port Metrics";

    json::Json data = json::Json();
    data[constants::EthernetSwitchPortMetrics::PACKETS] = json::Json::value_t::null;
    data[constants::EthernetSwitchPortMetrics::DROPPED_PACKETS] = json::Json::value_t::null;
    data[constants::EthernetSwitchPortMetrics::ERROR_PACKETS] = json::Json::value_t::null;
    data[constants::EthernetSwitchPortMetrics::BROADCAST_PACKETS] = json::Json::value_t::null;
    data[constants::EthernetSwitchPortMetrics::MULTICAST_PACKETS] = json::Json::value_t::null;
    data[constants::EthernetSwitchPortMetrics::ERRORS] = json::Json::value_t::null;
    data[constants::EthernetSwitchPortMetrics::BYTES] = json::Json::value_t::null;

    json::Json data_received = data;
    r[constants::EthernetSwitchPortMetrics::RECEIVED] = std::move(data_received);
    r[constants::EthernetSwitchPortMetrics::TRANSMITTED] = std::move(data);

    r[constants::EthernetSwitchPortMetrics::COLLISIONS] = json::Json::value_t::null;

    return r;
}

}


endpoint::EthernetSwitchPortMetrics::EthernetSwitchPortMetrics(const std::string& path) : EndpointBase(path) {}


endpoint::EthernetSwitchPortMetrics::~EthernetSwitchPortMetrics() {}


void endpoint::EthernetSwitchPortMetrics::get(const server::Request& req, server::Response& res) {
    auto r = ::make_prototype();
    r[Common::ODATA_ID] = PathBuilder(req).build();
    r[Common::ID] = constants::Common::METRICS;
    r[Common::NAME] = "Port " + req.params[PathParam::SWITCH_PORT_ID] + " Metrics";

    // check if port exists
    psme::rest::model::find<agent_framework::model::EthernetSwitch, agent_framework::model::EthernetSwitchPort>(
        req.params).get_uuid();

    set_response(res, r);
}

File: PODM/SW/pod-manager/podm-business-entities/src/main/java/com/intel/podm/business/entities/redfish/EthernetSwitchPortMetrics.java
/*
 * Copyright (c) 2017-2018 Intel Corporation
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.intel.podm.business.entities.redfish;

import com.intel.podm.business.entities.redfish.base.DiscoverableEntity;
import com.intel.podm.business.entities.redfish.base.Entity;
import com.intel.podm.business.entities.redfish.embeddables.SwitchPortMetrics;
import com.intel.podm.common.types.Id;

import javax.persistence.AttributeOverride;
import javax.persistence.AttributeOverrides;
import javax.persistence.Column;
import javax.persistence.Embedded;
import javax.persistence.Index;
import javax.persistence.JoinColumn;
import javax.persistence.OneToOne;
import javax.persistence.Table;
import java.util.Objects;

import static javax.persistence.CascadeType.MERGE;
import static javax.persistence.CascadeType.PERSIST;
import static javax.persistence.FetchType.LAZY;

@javax.persistence.Entity
@Table(name = "ethernet_switch_port_metrics", indexes = @Index(name = "idx_ethernet_switch_port_metrics_entity_id", columnList = "entity_id", unique = true))
public class EthernetSwitchPortMetrics extends DiscoverableEntity {
    @Column(name = "entity_id", columnDefinition = ENTITY_ID_STRING_COLUMN_DEFINITION)
    private Id entityId;

    @Embedded
    @AttributeOverrides({
        @AttributeOverride(name = "packets", column = @Column(name = "received_packets")),
        @AttributeOverride(name = "droppedPackets", column = @Column(name = "received_dropped_packets")),
        @AttributeOverride(name = "errorPackets", column = @Column(name = "received_error_packets")),
        @AttributeOverride(name = "broadcastPackets", column = @Column(name = "received_broadcast_packets")),
        @AttributeOverride(name = "multicastPackets", column = @Column(name = "received_multicast_packets")),
        @AttributeOverride(name = "bytes", column = @Column(name = "received_bytes")),
        @AttributeOverride(name = "errors", column = @Column(name = "received_errors"))
    })
    private SwitchPortMetrics received = new SwitchPortMetrics();

    @Embedded
    @AttributeOverrides({
        @AttributeOverride(name = "packets", column = @Column(name = "transmitted_packets")),
        @AttributeOverride(name = "droppedPackets", column = @Column(name = "transmitted_dropped_packets")),
        @AttributeOverride(name = "errorPackets", column = @Column(name = "transmitted_error_packets")),
        @AttributeOverride(name = "broadcastPackets", column = @Column(name = "transmitted_broadcast_packets")),
        @AttributeOverride(name = "multicastPackets", column = @Column(name = "transmitted_multicast_packets")),
        @AttributeOverride(name = "bytes", column = @Column(name = "transmitted_bytes")),
        @AttributeOverride(name = "errors", column = @Column(name = "transmitted_errors"))
    })
    private SwitchPortMetrics transmitted = new SwitchPortMetrics();

    @Column(name = "collisions")
    private Long collisions;

    @OneToOne(fetch = LAZY, cascade = {MERGE, PERSIST})
    @JoinColumn(name = "ethernet_switch_port_id")
    private EthernetSwitchPort ethernetSwitchPort;

    @Override
    public Id getId() {
        return entityId;
    }

    @Override
    public void setId(Id id) {
        this.entityId = id;
    }

    public SwitchPortMetrics getReceived() {
        return received;
    }

    public void setReceived(SwitchPortMetrics received) {
        this.received = received;
    }

    public SwitchPortMetrics getTransmitted() {
        return transmitted;
    }

    public void setTransmitted(SwitchPortMetrics transmitted) {
        this.transmitted = transmitted;
    }

    public Long getCollisions() {
        return collisions;
    }

    public void setCollisions(Long collisions) {
        this.collisions = collisions;
    }

    public EthernetSwitchPort getEthernetSwitchPort() {
        return ethernetSwitchPort;
    }

    public void setEthernetSwitchPort(EthernetSwitchPort ethernetSwitchPort) {
        if (!Objects.equals(this.ethernetSwitchPort, ethernetSwitchPort)) {
            unlinkEthernetSwitchPort(this.ethernetSwitchPort);
            this.ethernetSwitchPort = ethernetSwitchPort;
            if (ethernetSwitchPort != null && !this.equals(ethernetSwitchPort.getMetrics())) {
                ethernetSwitchPort.setMetrics(this);
            }
        }
    }

    public void unlinkEthernetSwitchPort(EthernetSwitchPort ethernetSwitchPort) {
        if (Objects.equals(this.ethernetSwitchPort, ethernetSwitchPort)) {
            this.ethernetSwitchPort = null;
            if (ethernetSwitchPort != null) {
                ethernetSwitchPort.unlinkMetrics(this);
            }
        }
    }

    @Override
    public void preRemove() {
        unlinkEthernetSwitchPort(ethernetSwitchPort);
    }

    @Override
    public boolean containedBy(Entity possibleParent) {
        return isContainedBy(possibleParent, ethernetSwitchPort);
    }
}
